OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+  using ColLabelsIterPairType      = std::pair<ColLabelsIterType,
+                                               ColLabelsIterType>;
+  using ColLabelsConstIterPairType = std::pair<ColLabelsConstIterType,
+                                               ColLabelsConstIterType>;
+
+  /// Construct empty DataTable.
+  DataTable_() :
+    data{}, metadata{}, col_ind{} {}
+
+  /// Construct DataTable of given shape populated with given value val. 
+  /// Default value for val is NaN.
+  /// \param nrows Number of rows.
+  /// \param ncols Number of columns.
+  /// \param val   Value to initialize all the entries to.
+  DataTable_(size_t nrows,
+             size_t ncols,
+             const ET& val = ET{SimTK::NaN}) :
+    data{int(nrows), int(ncols), val}, metadata{}, col_ind{} {}
+
+  /// Construct DataTable using an iterator(satisfying requirement of an 
+  /// input_iterator) which produces one entry at a time. The entries of 
+  /// DataTable_ are copy initialized using the values produced by the 
+  /// iterator. For example, specifying ROWWISE for parameter dir and 10 for 
+  /// parameter ndir will populate the DataTable one row at a time with each 
+  /// row's length taken to be 10 elements.
+  ///
+  /// \param first Beginning of range covered by the iterator.
+  /// \param last End of the range covered by the iterator.
+  /// \param ndir Extent of the dimention specified by parameter dir. 
+  /// \param dir Dimension to populate the DataTable. Populate row-wise or 
+  ///              column wise. See OpenSim::InputItDir for possible values.
+  /// \param allow_missing Allow for missing values. When set to false, this
+  ///                      function will throw if the input iterator fills up 
+  ///                      the last row/col partially. When set to true, 
+  ///                      missing elements are set to SimTK::NaN.
+  ///
+  /// \throws ZeroElements When input-iterator does not produce any elements.
+  ///                      That is first == last.
+  template<typename InputIt>
+  DataTable_(InputIt first,
+             typename std::enable_if<!std::is_integral<InputIt>::value,
+                                     InputIt>::type last,
+             const size_t ndir,
+             const InputItDir dir = ROWWISE,
+             const bool allow_missing = false) :
+    data{static_cast<int>(dir == ROWWISE ? 1    : ndir), 
+         static_cast<int>(dir == ROWWISE ? ndir :    1)},
+    metadata{},
+    col_ind{} {
+    if(!(first != last))
+      throw ZeroElements{"Input iterators produce zero elements."};
+    if(ndir == 0)
+      throw InvalidEntry{"Input argument 'ndir' cannot be zero."};
+
+    int row{0};
+    int col{0};
+    while(first != last) {
+      data.set(row, col, *first);
+      ++first;
+      if(dir == ROWWISE) {
+        ++col;
+        if(col == static_cast<int>(ndir)  && first != last) {
+          col = 0;
+          ++row;
+          data.resizeKeep(data.nrow() + 1, data.ncol());
```

Comment:
> This is similar to what std::vector does. It re-allocates and re-locates everytime we hit the capacity.

std::vector is required to double the capacity each time it runs out so that the average performance is O(n). That's easy to do in 1D but it isn't as obvious to me what should be done in 2D. Currently Matrix doesn't do any preallocation. Probably the best fix would be to make resizeKeep() smarter; that would take some work but we could make that part of this project. I filed a Simbody issue simbody/simbody#401.